Clean single sample microsalt cases

GustavLagneborgScilifelab opened the pull request:
## Description

### Added
- A function that checks the file age for microsalt run dirs. This functionality allows for cleaning single sample cases aswell

### Changed

-

### Fixed

-


### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

henrikstranneheim commented on cg/meta/workflow/microsalt.py:
```diff
@@ -64,15 +64,30 @@ def process(self) -> Process:
+        return case_paths
+
+    def verify_case_path_age(self, case_paths: List[Path], case_id: str):
+        old_date = datetime.now() - timedelta(days=21)
```
Type hint

henrikstranneheim replied:
```suggestion
        due_date = datetime.now() - timedelta(days=ANALYSIS_DUE_DATE)
```
Make 21 a parameter into the functiion to make it more general and call it using a constant instead of `21`

henrikstranneheim commented on cg/meta/workflow/microsalt.py:
```diff
@@ -64,15 +64,30 @@ def process(self) -> Process:
+    def verify_case_path_age(self, case_paths: List[Path], case_id: str):
+        old_date = datetime.now() - timedelta(days=21)
+        for case in case_paths:
+            creation_date = datetime.fromtimestamp(os.path.getctime(case))
```
type hint

henrikstranneheim commented on cg/meta/workflow/microsalt.py:
```diff
@@ -64,15 +64,30 @@ def process(self) -> Process:
+
+        return case_paths
+
+    def verify_case_path_age(self, case_paths: List[Path], case_id: str):
```
```suggestion
    def verify_case_paths_age(self, case_paths: List[Path], case_id: str):
```

henrikstranneheim reviewed (commented):
👍

GustavLagneborgScilifelab commented:
### Dry run with command `cg clean microsalt-past-run-dirs -d -y "2022-11-30"`:
Case happyboar should be skipped as it has younger files than 21 days, and case luckyteal should delete two paths (one single sample and one with all samples). I created those 2 paths manually and then modified the m-time for them with the touch command (`touch -a -m -t "201203101513" ACC9964_2022.11.13_10.43.35/`)

```Shell
Cleaning 2 analyses created before 2022-11-30 00:00:00
Cleaning microsalt output for happyboar
Case happyboar exists in status db
All paths in happyboar is not older than 21 days, skipping and going to next case!
Cleaning microsalt output for luckyteal
Case luckyteal exists in status db
Would have deleted: [PosixPath('/home/proj/stage/microbial/results/ACC9964A1_2022.11.13_10.43.35'), PosixPath('/home/proj/stage/microbial/results/ACC9964_2022.11.13_10.43.35')]
Done cleaning microsalt output
```

GustavLagneborgScilifelab commented:
### Real run with command: `cg clean microsalt-past-run-dirs -y "2022-11-30"`
Run dir folder before:
![Screenshot 2022-11-25 at 16 00 15](https://user-images.githubusercontent.com/107848458/204011533-d5a35540-b419-4587-b53e-2c167d482ee8.png)
Output of command:
```Shell
Cleaning 2 analyses created before 2022-11-30 00:00:00
Cleaning microsalt output for happyboar
Case happyboar exists in status db
All paths in happyboar is not older than 21 days, skipping and going to next case!
Cleaning microsalt output for luckyteal
Case luckyteal exists in status db
Cleaned /home/proj/stage/microbial/results/ACC9964A1_2022.11.13_10.43.35
Cleaned /home/proj/stage/microbial/results/ACC9964_2022.11.13_10.43.35
Adding a cleaned at date for case luckyteal
Done cleaning microsalt output
````
Run dir after command:
![Screenshot 2022-11-25 at 16 02 20](https://user-images.githubusercontent.com/107848458/204011704-8801d08c-4e5b-43ce-af8b-ed109a685359.png)

The expected outcome is that both single sample and all sample analyses for case luckyteal (ACC9964) should be deleted

karlnyr reviewed (commented):
🔥

GustavLagneborgScilifelab commented on tests/cli/clean/test_microbial_clean.py:
```diff
@@ -107,3 +110,20 @@ def test_clean_run(
+    caplog.set_level(logging.INFO)
+    run_dir = [tmpdir_factory.mktemp("microsalt_case")]
+
+    with pytest.raises(FileNotFoundError):
+        # WHEN at least one of the paths is younger than 21 days
+        clean_context_microsalt.meta_apis["analysis_api"].verify_case_paths_age(
+            run_dir, microsalt_case_clean
+        )
+
+    # THEN a FileNotFoundError should be raised and logged in the log
+    assert "is not older than 21 days, skipping and going to next case!" in caplog.text
```
Agreed!

karlnyr commented on tests/cli/clean/test_microbial_clean.py:
```diff
@@ -2,10 +2,13 @@
 from cg.cli.workflow.commands import clean_run_dir
 from cg.constants import Pipeline
 from click.testing import CliRunner
+from tests.conftest import microsalt_dir
```
I dont think you need to import this? Or do you have another microsalt_dir fixture lower down in the fixture tree?

karlnyr reviewed (approved):
minor comments!

GustavLagneborgScilifelab commented:
Installed in production:
<img width="1239" alt="Screenshot 2022-11-28 at 13 22 08" src="https://user-images.githubusercontent.com/107848458/204276733-39cda28e-9daf-4cd2-ae35-3d97862f7731.png">
